Mathlib/Topology/Algebra/Module/PerfectPairing.lean

```diff
@@ -0,0 +1,70 @@
+/-
+Copyright (c) 2025 Yaël Dillies, Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Andrew Yang
+-/
+import Mathlib.LinearAlgebra.BilinearMap
+import Mathlib.Topology.Algebra.Module.LinearMap
+
+/-!
+# Continuous perfect pairings
+
+This file defines continuous perfect pairings.
+
+For a topological ring `R` and two topological modules `M` and `N`, a continuous perfect pairing is
+a continuous bilinear map `M × N → R` that is bijective in both arguments.
+
+Note that this does **not** imply that the continuous dual of `M` is `N` and vice-versa (one would
+instead need "homeomorphic in both arguments").
+
+## TODO
+
+Adapt `PerfectPairing` to this Prop-valued typeclass paradigm
+-/
+
+open Function
+
+namespace LinearMap
+variable {R M N : Type*}
+  [CommRing R] [TopologicalSpace R] [AddCommGroup M] [Module R M] [TopologicalSpace M]
+  [AddCommGroup N] [Module R N] [TopologicalSpace N] (p : M →ₗ[R] N →ₗ[R] R) {x : M} {y : N}
+
+/-- For a topological ring `R` and two topological modules `M` and `N`, a continuous perfect pairing
+is a continuous bilinear map `M × N → R` that is bijective in both arguments.
+
+Note that this does **not** imply that the dual of `M` is `N` and vice-versa (one would instead need
+"homeomorphic in both arguments"). -/
+@[ext]
+class IsContPerfPair (p : M →ₗ[R] N →ₗ[R] R) where
```

Comment:
Are you sure you want `class` and not just `structure`?

Obviously both will work but I'm generally a bit slower to use typeclasses when the keys are not just raw types: in this case `p` is data which could be rewritten or quite complex. It's easy to imagine the user having to start building would-be instances by hand in a complex argument.

Reply:
Yes, I am pretty sure.

As Anatole points out, we might want in the future to have a whole hierarchy of "continuous perfect pairings", so having to pass the instance around by hand might become unwieldy.

On the other hand, having now worked somewhat extensively with typeclasses-on-terms in the context of category theory, I can confirm that it works pretty well and only occasionally do you need to massage an existing assumption to a syntactically-different-but-defeq form for typeclass search to find it. In our case, it looks particularly innocuous since the typeclass is Prop-valued.